### DISCUSSION TITLE
Reworking DecryptionStream and OpenPgpMetadata

### DISCUSSION THREAD

**Discussion by vanitasvitae** [pgpainless/pgpainless#317] [Q&A]
Hey!
I'm currently working on a rework of the way PGPainless decrypts messages.
While the current implementation works mostly fine, it has the drawback not to reject some malformed messages (see #237 for more details).

Currently I am working on a new OpenPpgMessageInputStream which enforces the OpenPGP message format described in section 11.3 of RFC4880.
That made me wondering: How do downstream users of PGPainless use the OpenPgpMetadata class?

I'm asking because I noticed that the current implementation does not reflect the capabilities of OpenPGP really well.
Most notably, a valid OpenPGP message *may* contain multiple layers of encryption and compression, so having a `getSymmetricKeyAlgorithm()` method returning a single algorithm does not really make sense. The same is true for `getCompressionAlgorithm()`.

To be fair, a single encryption-layer plus a single compression layer is what makes up 99% of (encrypted) OpenPGP messages, so I'm probably concerned with a corner-case, yet the current API is not perfect either.

Would you mind if I change the `getSymmetricKeyAlgorithm()`, `getCompressionAlgorithm()`, `getSessionKey()` methods to return `Iterator` objects instead? That way we could combine the algorithms of the several layers and provide access to them for the user.
Can you think of a better solution? Perhaps some kind of result consisting of nested objects which resemble the message structure?

@tomholub @DenBond7 @IvanPizhenko @msfjarvis

**Comment by msfjarvis** [comment]
AFAIK we do not use these APIs in Password Store, but from what I'm reading it looks like there's a solid argument in favour of the change to Iterators so it has my vote.

**Comment by IvanPizhenko** [comment]
We also seem to not use these APIs, so I assume it will have no effect on us.

**Comment by tomholub** [comment]
@DenBond7 to double check on Android

**Reply by DenBond7** [reply]
> @DenBond7 to double check on Android

We also don't use these APIs on Android yet.
